Sources/Caching/PostedTransactionCache.swift

```diff
@@ -19,28 +19,78 @@ protocol PostedTransactionCacheType: Sendable {
     func savePostedTransaction(_ transaction: StoreTransactionType)
     func hasPostedTransaction(_ transaction: StoreTransactionType) -> Bool
 
+    /// - Returns: the subset of `transactions` that have not been posted.
+    func unpostedTransactions<T: StoreTransactionType>(in transactions: [T]) -> [T]
+
 }
 
 final class PostedTransactionCache: PostedTransactionCacheType {
 
-    private typealias StoredTransactions = Set<String>
+    /// The format of the cache: `StoreTransactionType.transactionIdentifier` -> `Date`.
+    private typealias StoredTransactions = [String: Date]
 
     private let deviceCache: DeviceCache
+    private let clock: ClockType
 
-    init(deviceCache: DeviceCache) {
+    init(deviceCache: DeviceCache, clock: ClockType) {
         self.deviceCache = deviceCache
+        self.clock = clock
     }
 
     func savePostedTransaction(_ transaction: StoreTransactionType) {
+        RCIntegrationTestAssertNotMainThread()
+
+        Logger.debug(Strings.purchase.saving_posted_transaction(transaction))
+
         self.deviceCache.update(key: CacheKey.transactions,
-                                default: Set<String>()) { transactions in
-            transactions.insert(transaction.transactionIdentifier)
+                                default: StoredTransactions()) { transactions in
+            self.pruneOldTransactions(from: &transactions)
+            transactions[transaction.transactionIdentifier] = self.clock.now
```

Comment:
This entire modification is done _in place_ so we don't end up with 2 dictionary copies in memory.